swift/extractor/SwiftDispatcher.h

```diff
@@ -57,18 +21,21 @@ class SwiftDispatcher {
   SwiftDispatcher(const swift::SourceManager& sourceManager, TrapArena& arena, TrapOutput& trap)
       : sourceManager{sourceManager}, arena{arena}, trap{trap} {}
 
+  template <typename Entry>
+  void emit(const Entry& entry) {
+    trap.emit(entry);
+  }
+
   // This is a helper method to emit TRAP entries for AST nodes that we don't fully support yet.
-  template <typename Parent, typename Child>
-  void TBD(Child* entity, const std::string& suffix) {
-    using namespace std::string_literals;
+  template <typename E>
+  void emitUnknown(E* entity) {
     auto label = assignNewLabel(entity);
-    auto kind = detail::getKindName<Parent>(static_cast<const Parent*>(entity)->getKind());
-    auto name = "TBD ("s + kind + suffix + ")";
-    if constexpr (std::is_same_v<Parent, swift::TypeBase>) {
-      trap.emit(UnknownTypesTrap{label, name});
-    } else {
-      trap.emit(UnknownAstNodesTrap{label, name});
-    }
+    using Trap = BindingTrapOf<E>;
+    static_assert(sizeof(Trap) == sizeof(label),
```

Comment:
`ToBindingTrapFunctor` seems pretty empty to me, so doesn't this always hold? I'm probably just missing details here.

Reply:
`ToBindingTrapFunctor` is implemented by `TrapEntries.h` in the [generated headers](https://github.com/github/codeql/suites/6588171517/artifacts/247105982). This particular assertion fires whenever a final class in `schema.yml` has a single property (i.e. non-optional, non-repeated). For example uncommenting line 995 in `schema.yml` (field `introducer_int` in class `ConcreteVarDecl`) will result in 
```
./swift/extractor/SwiftDispatcher.h:34:5: error: static_assert failed due to requirement 'sizeof(codeql::ConcreteVarDeclsTrap) == sizeof (label)' "Binding traps of unknown entities must only have the `id` field (the class should be empty in schema.yml)"
    static_assert(sizeof(Trap) == sizeof(label),
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
I still don't get this, because:
```
template<>
struct ToBindingTrapFunctor<TrapLabel<FileTag>> {
  using type = FilesTrap;
};
```
pretty much looks like an empty class to me.

Reply:
ah, sorry, I did not get that that was your doubt.

It's the `type` nested alias that does the trick. From `TrapTagTraits`:
```cpp
template <typename T>
using BindingTrapOf = typename detail::ToBindingTrapFunctor<TrapLabelOf<T>>::type;
```
so for example `BindingTrapOf<swift::VarDecl>` will resolbe to `typename detail::ToBindingTrapFunctor<TrapLabel<ConcreteVarDeclTag>>::type` which in the generated file is defined as `ConcreteVarDeclsTrap`.

Reply:
This is the same pattern used for transformation type traits like [`std::underlying_type`](https://en.cppreference.com/w/cpp/types/underlying_type) and similar transformations in `type_traits`. Although I see I am using mixed naming conventions using lowercase `type`. The `std` convention would have two separate definitions like in
```cpp
template <typename T>
using binding_trap_of = detail::to_binding_trap_functor<...>;

template <typename T>
using binding_trap_of_t = typename binding_trap_of<T>::type;
```

Reply:
Sure but that's because there's actual work needed to implement `underlying_type` it's not just an alias like we have here.

Reply:
> Got it. But then why do this at all and not just:
> 
> ```
> static_assert(sizeof(E) == sizeof(label),
> ```

`E` is the swift frontend tool entity (like `swift::VarDecl`) which is distinct from the trap entry (for example `ConcreteVarDeclsTrap`). We also need to compute that trap entry class to actually emit it on [line 37 below](https://github.com/github/codeql/blob/redsun82/swift-tbd-rework/swift/extractor/SwiftDispatcher.h#L37)

Reply:
Yes, that should probably just be `TrapLabel<E>{label}`.?

Reply:
> Sure but that's because there's actual work needed to implement `underlying_type` it's not just an alias like we have here.

I think that alias is there to hide away an implementation detail, which does the actual work in terms of template specialization. 

Would you find it clearer if we wrote something like
```cpp
  template <typename Label>
  using BindingTrapOf = typename detail::ToBindingTrap<Label>::type;
```
and then in the code using it did:
```cpp
  using Trap = BindingTrapOf<decltype(label)>;
```
or more explicitly
```
  using Label = TrapLabelOf<E>;
  using Trap = BindingTrapOf<Label>;
  Label label = assignNewLabel(e);
  ...
```
?

Reply:
I understand less and less of what this is code is doing. Why is all this indirect mess needed at all? This is going to cause a lot of trouble for anyone who is going to maintain this in the future.

Reply:
I want something that's just 2 lines of code instead of 12.

Reply:
> Yes, that should probably just be `TrapLabel<E>{label}`.?

Hmm, no, labels cannot be emitted in place of trap entries. `TrapLabelOf<E>` which resolves to `TrapLabel<TrapTagOf<E>>` (for example, `TrapLabel<ConcreteVarDeclTag>`) denotes a typed trap label, so something that is a `#` label in the trap file. We need to emit an actual trap (like `concrete_var_decls(#1e)`) which in C++ is done printing a `ConcreteVarDeclsTrap` class.